### PR TITLE
Feature 55319 ose secret sti

### PR DESCRIPTION
**Pull request by TheWalkingLeek**

### CODE REVIEW

File: db/migrate/20200917140436_add_type_to_account.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class AddTypeAndEncryptedDataToAccount < ActiveRecord::Migration[6.0]
+  def up
+    add_column :accounts, :type, :string, default: 'Account::Credentials', null: false
+    change_column :accounts, :type, :string, default: nil
```
Comment (psunix) [inline db/migrate/20200917140436_add_type_to_account.rb:4]: maybe we could just keep the default value? 
still thinking about anything that could go wrong if the type attr has a default value ... nothing coming up so far :joy_cat:

**Review by psunix — CHANGES_REQUESTED** [review]
looks pretty good, lot's of code !

the only no-go is the use of Account specific details inside CryptUtils. That's really bad design like this.

File: app/models/account/ose_secret.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class Account::OSESecret < Account
+  attr_accessor :data
+
+  def decrypt(team_password)
+    @data = decrypt_data(team_password)
+  end
+
+  def encrypt(team_password)
```
[the remark this answers is not in the document]
Comment (TheWalkingLeek) [reply app/models/account/ose_secret.rb:13]: This isn't possible as far as I can see it. Problem is, that you have to assign the encrypted_data field to the Account entry when encrypting, and the ose_secret field when decrypting. You can see how I adjusted it in the review changes :)

File: app/models/account/encrypted_data.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class Account::EncryptedData
```
Comment (psunix) [inline app/models/account/encrypted_data.rb:3]: how about:

cleartext_data, data, iv

#encrypt(encryption_key), #decrypt(encryption_key)

#to_json implement your-self (returns the correct attrs, raises exception if #encrypt was not called -> data is empty)

File: app/models/account/ose_secret.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (psunix) [reply app/models/account/ose_secret.rb:13]: you're right !

File: lib/crypt_utils.rb [hunk 4]
```diff
@@ -109,6 +109,30 @@ def decrypt_blob(blob, team_password)
       decrypted_blob << cipher.final
       decrypted_blob.force_encoding('UTF-8')
     end
-  end
 
+    def encrypt_base64(data, cipher_key, iv)
```
Comment (psunix) [inline lib/crypt_utils.rb:113]: here you called it cipher_key ... further up, it's called encryption_key. 

I prefer encryption_key over cipher_key ... or just key ?

**Review by psunix — CHANGES_REQUESTED** [review]
magnifique !!!

some small things to take care of.

**Review by psunix — APPROVED** [review]
:rocket: :partying_face:

**Pull request merged**